sympy/core/sympify.py

```diff
@@ -263,19 +263,16 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
 
     """
     try:
-        if a in sympy_classes:
+        if a.__sympy__:
             return a
-    except TypeError: # Type of a is unhashable
+    except AttributeError:  # a is not a sympy object
         pass
-    cls = getattr(a, "__class__", None)
-    if cls is None:
-        cls = type(a) # Probably an old-style class
-    if cls in sympy_classes:
-        return a
 
     if isinstance(a, CantSympify):
         raise SympifyError(a)
-
+    cls = getattr(a, "__class__", None)
+    if cls is None:
+        cls = type(a)  # Probably an old-style class
     try:
         return converter[cls](a)
     except KeyError:
```

Comment:
While you're here that KeyError is being caught to broadly. It should have been written something like:
```
    try:
        c = converter[cls]
    except KeyError:
        # whatever goes here
    else:
        return c(a) # Don't mask KeyErrors leaking from this function call
```
Better still though don't catch KeyError (use get instead):
```
    c = converter.get(cls, None)
    if c is None:
```